src/model.rs

```diff
@@ -47,45 +49,115 @@ impl Marginal {
+
+                        //achtung: the distance matrix should be gotten rid of the distance 0 entries, otherwise this will not work.
+                        //todo: double check that this works as expected.
+                        //define x
+                        let x: usize = 2;
+
+                        //find the current haplotype
+                        let current_haplotype = &self.haplotypes[haplotype_index];
+                        dbg!(&current_haplotype);
 
-                    //find the index of the (haplotype, haplotype_group) in graph
-                    let index = &self
-                        .haplotype_graph
-                        .node_indices()
-                        .find(|i| {
-                            &self.haplotype_graph[*i]
-                                == &(current_haplotype.clone(), haplotype_group.clone())
-                        })
-                        .unwrap();
-
-                    // step through the graph and sum incoming edges into the node weight
-                    let mut bfs = Bfs::new(&self.haplotype_graph, *index);
-
-                    while let Some(nx) = bfs.next(&self.haplotype_graph) {
-                        // we can access `graph` mutably here still
-                        let haplotype_query = &self.haplotype_graph[nx].0;
-                        for (h, f) in self.haplotypes[0..haplotype_index]
-                            .to_vec()
-                            .iter()
-                            .zip(fractions[0..haplotype_index].to_vec().iter())
-                        {
-                            if (h == haplotype_query) && (f > &NotNan::new(0.0).unwrap()) {
+                        //loop over haplotypes in the distance matrix and find haplotypes that are at distance x to the current haplotype (similar_L(h))
+                        if let Some(distance_matrix) = &self.distance_matrix {
+                            dbg!(&distance_matrix);
+                            dbg!(&self.lp_haplotypes);
+
+                            let mut lp_haplotype_list = self.lp_haplotypes.clone().unwrap();
+
+                            //retain all elements in lp haplotypes except the current haplotype
+                            lp_haplotype_list.retain(|x| x != current_haplotype);
+
+                            //compute similar_L(current_haplotype)
+                            let mut similar_l = 0;
+
+                            for h in lp_haplotype_list.iter() {
+                                for ((h1, h2), distance) in distance_matrix.iter() {
+                                    if ((h1 == h && h2 == current_haplotype)
+                                        || (h2 == h && h1 == current_haplotype))
+                                        && (*distance < x)
+                                    {
+                                        similar_l += 1;
+                                    }
+                                }
+                            }
```

Comment:
This is very slow inside the recursion. First a copy of the haplotypes is made (in every recursion step). Then you filter it with retain (linear runtime again). Just to iterate over it below. I would rather determine the similar haplotypes before going into calculating the marginal, so that you don't need to recompute them all the time in the recursion.

Reply:
You're right, I only focused on making this to work initially.

Reply:
This is improved now.